Fork aware primary block processing

liuchengxu opened the pull request:
This PR aims to handle the primary block forks properly. Basically, one incoming primary block import notification could mean multiple primary blocks we need to process in case the primary chain forks.

Tested in a local network with multiple farmers, now the domain forks whenever the primary chain forks.

```
2022-12-23 00:41:36.915  INFO tokio-runtime-worker sc_informant: [SecondaryChain] ♻️  Reorg on #555,0x7e2d…c18a to #555,0x5e3e…d1f9, common ancestor #554,0x593d…9d31    
2022-12-23 00:41:38.504  INFO tokio-runtime-worker sc_informant: [PrimaryChain] ♻️  Reorg on #557,0xc0f0…68e2 to #557,0x94b8…d90e, common ancestor #556,0x787e…3d28    
2022-12-23 00:41:38.506  INFO tokio-runtime-worker sc_informant: [CoreDomain] ♻️  Reorg on #557,0xe471…8a27 to #557,0xa594…fd25, common ancestor #556,0x77ab…8c41    
2022-12-23 00:41:38.508  INFO tokio-runtime-worker sc_informant: [SecondaryChain] ♻️  Reorg on #557,0xd008…fa6c to #557,0x11be…01dd, common ancestor #556,0x59a1…0cca    
2022-12-23 00:41:44.321  INFO tokio-runtime-worker sc_informant: [PrimaryChain] ♻️  Reorg on #560,0x1afc…6ad7 to #560,0xbbcd…1f74, common ancestor #559,0xfd68…a546    
2022-12-23 00:41:44.360  INFO tokio-runtime-worker sc_informant: [SecondaryChain] ♻️  Reorg on #560,0x0976…b034 to #560,0xe161…3528, common ancestor #559,0x4659…c933    
2022-12-23 00:41:44.375  INFO tokio-runtime-worker sc_informant: [CoreDomain] ♻️  Reorg on #560,0xa8b0…3a23 to #560,0x7656…f0e1, common ancestor #559,0x488b…4a5f    
2022-12-23 00:41:44.649  INFO tokio-runtime-worker sc_informant: [PrimaryChain] ♻️  Reorg on #560,0xbbcd…1f74 to #561,0x9668…1737, common ancestor #559,0xfd68…a546    
2022-12-23 00:41:44.679  INFO tokio-runtime-worker sc_informant: [SecondaryChain] ♻️  Reorg on #560,0xe161…3528 to #561,0xd291…3f15, common ancestor #559,0x4659…c933    
2022-12-23 00:41:48.664  INFO tokio-runtime-worker sc_informant: [PrimaryChain] ♻️  Reorg on #562,0x1846…f0bf to #562,0x756d…d3cb, common ancestor #561,0x9668…1737    
2022-12-23 00:41:48.682  INFO tokio-runtime-worker sc_informant: [SecondaryChain] ♻️  Reorg on #562,0x74e9…4617 to #562,0x7d05…5784, common ancestor #561,0xd291…3f15    
2022-12-23 00:41:48.693  INFO tokio-runtime-worker sc_informant: [CoreDomain] ♻️  Reorg on #562,0x11e4…e3ab to #562,0x5cda…248d, common ancestor #561,0xfa16…214a    
2022-12-23 00:41:48.996  INFO tokio-runtime-worker sc_informant: [PrimaryChain] ♻️  Reorg on #563,0x41db…089a to #563,0x951a…e429, common ancestor #561,0x9668…1737    
2022-12-23 00:41:49.026  INFO tokio-runtime-worker sc_informant: [SecondaryChain] ♻️  Reorg on #563,0xc445…8871 to #563,0xf00b…f148, common ancestor #561,0xd291…3f15    
2022-12-23 00:42:06.215  INFO tokio-runtime-worker sc_informant: [PrimaryChain] ♻️  Reorg on #575,0xdc48…bfe1 to #575,0x0497…ea24, common ancestor #574,0x4785…4fb5    
2022-12-23 00:42:06.225  INFO tokio-runtime-worker sc_informant: [SecondaryChain] ♻️  Reorg on #575,0x72cf…6ce7 to #575,0xcb82…af8c, common ancestor #574,0x77a7…c29a    
2022-12-23 00:42:06.227  INFO tokio-runtime-worker sc_informant: [CoreDomain] ♻️  Reorg on #575,0x7bd5…251a to #575,0x1db0…5381, common ancestor #574,0xb45e…057b    
2022-12-23 00:42:20.069  INFO tokio-runtime-worker sc_informant: [PrimaryChain] ♻️  Reorg on #584,0x060d…03a9 to #584,0x8871…f48a, common ancestor #583,0xb522…4b91    
2022-12-23 00:42:20.087  INFO tokio-runtime-worker sc_informant: [SecondaryChain] ♻️  Reorg on #584,0x8d1d…8a5c to #584,0xb41e…21d0, common ancestor #583,0x7fa5…d4f1    
2022-12-23 00:42:20.405  INFO tokio-runtime-worker sc_informant: [PrimaryChain] ♻️  Reorg on #586,0x1e04…5650 to #586,0xa8d9…b9a8, common ancestor #585,0x9d06…1dd7    
2022-12-23 00:42:20.408  INFO tokio-runtime-worker sc_informant: [CoreDomain] ♻️  Reorg on #586,0x4da2…3115 to #586,0xcd18…eaa0, common ancestor #585,0x3abf…49f6    
2022-12-23 00:42:20.408  INFO tokio-runtime-worker sc_informant: [SecondaryChain] ♻️  Reorg on #586,0xcc21…7b58 to #586,0x3ca7…d3fa, common ancestor #585,0xacf2…7f6d    
2022-12-23 00:42:26.066  INFO tokio-runtime-worker sc_informant: [PrimaryChain] ♻️  Reorg on #590,0xf57b…f879 to #590,0xe23c…51af, common ancestor #589,0x79fd…e6bd    
2022-12-23 00:42:26.079  INFO tokio-runtime-worker sc_informant: [CoreDomain] ♻️  Reorg on #590,0xc052…32c5 to #590,0x67df…48c0, common ancestor #589,0x9b5c…d083    
2022-12-23 00:42:26.080  INFO tokio-runtime-worker sc_informant: [SecondaryChain] ♻️  Reorg on #590,0x6b47…3b77 to #590,0xed6b…6da2, common ancestor #589,0xea22…9d03    
2022-12-23 00:42:27.422  INFO tokio-runtime-worker sc_informant: [PrimaryChain] ♻️  Reorg on #591,0x7385…0415 to #591,0xf609…b6f1, common ancestor #589,0x79fd…e6bd    
2022-12-23 00:42:27.454  INFO tokio-runtime-worker sc_informant: [CoreDomain] ♻️  Reorg on #591,0xb923…83a8 to #591,0xd6a1…a7b0, common ancestor #589,0x9b5c…d083    
2022-12-23 00:42:27.473  INFO tokio-runtime-worker sc_informant: [SecondaryChain] ♻️  Reorg on #591,0xe782…e23f to #591,0x879b…c623, common ancestor #589,0xea22…9d03    
```

Close #1034

This is not the whole story, I still see some edge cases locally and need more time to test it both locally and on gemini-3b, but the meat is there, I hope it can be reviewed sooner than later.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

liuchengxu committed:
Extract `preprocess_primary_block()` in domain_block_processor module

No logical changes.

liuchengxu committed:
Move the primary block preprocessing into bundle processor instead of passing the preprocessed result

liuchengxu committed:
Remove helpless `process_primary_block()` in domain_worker

liuchengxu committed:
Extract `process_bundles_at()` in bundle processor

Given a primary block, the parent domain block used to build the new
domain block corresponding to the given primary block should be deterministic.

liuchengxu committed:
Handle the primary block import in a fork-aware way

In short, the incoming imported primary block can be on a fork and we
need to deal with that on our own instead of having a misunderstanding
that it will handled automatically in the low level.

liuchengxu committed:
Disable an assertion in core_bundle_processor

liuchengxu committed:
Ensure the new domain block number matches the corresponding primary block number

Something is wrong if they mismatches.

nazar-pc commented on crates/subspace-fraud-proof/src/tests.rs:
```diff
@@ -122,7 +122,7 @@ async fn execution_proof_creation_and_verification_should_work() {
     };
 
-    let bundles = vec![OpaqueBundle {
+    let _bundles = vec![OpaqueBundle {
```
Why do we still need it here?

nazar-pc commented on domains/client/domain-executor/src/core_bundle_processor.rs:
```diff
@@ -115,11 +121,41 @@ where
+        for (i, primary_info) in primary_imports.iter().enumerate() {
+            // Use the origin fork_choice for the target primary block,
+            // the intermediate ones use `Custom(false)`.
+            let fork_choice = if i == primary_imports.len() - 1 {
```
I have a habit to think if index management can be avoided, looks like it can be here:
```suggestion
            let fork_choice = if primary_info == primary_imports.last() {
```

liuchengxu replied:
Can't be used like this as `HashAndNumber` does not derive `PartialEq`.

rahulksnv replied:
Sorry to reopen this, but would be great to avoid the index check. Re: `primary_info == primary_imports.last()` check: can lead to false positives if there are duplicates, if the last element happens to be present early on in the iteration (not sure this applies in this case) 

Other solution would be use `peekable()`, something like:

```
let mut iter = primary_imports.iter().peekable();
while let Some(primary_info) = iter.next() {
    let fork_choice = iter.peek().map_or(ForkChoiceStrategy::Custom(false), || fork_choice);
    // ...
}
```

nazar-pc replied:
There will be no duplicates here, but peekable looks nicer indeed, I always forget about is

nazar-pc commented on domains/client/domain-executor/src/domain_block_processor.rs:
```diff
@@ -182,6 +196,130 @@ where
         }
     }
 
+    pub(crate) fn fork_aware_block_imports(
```
Would be helpful to have some comments on this function. It doesn't seem to import anything, so maybe a more descriptive name as well.

liuchengxu replied:
Refactor it a bit, let me know if you have more questions or suggestions.

nazar-pc commented on domains/client/domain-executor/src/domain_block_processor.rs:
```diff
@@ -182,6 +196,130 @@ where
+            "Calculating ForkAwareBlockImports on #{best_number},{best_hash:?}"
+        );
+
+        match (retracted.is_empty(), enacted.is_empty()) {
```
I'd use nested `if/else` here for readability. Yes, it will add one more layer of indentation, but I wouldn't need to scroll up to see what two booleans mean.

liuchengxu replied:
The code was initially written when you don't have enough knowledge about the underlying details (even you thought you do, it's possible to have some misunderstanding somewhere), using `match` forces you to list all the possible cases and handle them or give an explanation, `retracted` and `enacted` combined here helps me understand the fork scenario in the real world so that you can think them from a bigger context instead of branching into a smaller context each time.

rahulksnv replied:
I also had the issue where I had to keep moving up and down. One alternative is to use plain if instead, something like:

```
if retracted.is_empty() && enacted.is_empty() {
   // ...
} else if !retracted.is_empty() && enacted.is_empty() { 
   // ...
} else ...
```

nazar-pc commented on domains/client/domain-executor/src/domain_block_processor.rs:
```diff
@@ -182,6 +196,130 @@ where
+                })
+            }
+            (true, true) => {
+                unreachable!("Tree route must not be empty as `from` and `to` in tree_route() must not be the same.");
```
This looks more like a proof:
```suggestion
                unreachable!("Tree route is not empty because `primary_from` and `primary_to` in tree_route() are not be the same.");
```

nazar-pc commented on domains/client/domain-executor/src/domain_block_processor.rs:
```diff
@@ -182,6 +196,130 @@ where
+            }
+            (false, false) => {
+                match retracted.len().cmp(&enacted.len()) {
+                    Ordering::Less | Ordering::Equal => {
```
What about this?:
```suggestion
                if retracted.len() <= enacted.len() {
```

liuchengxu replied:
I can take this as all cases of this branch are listed and can be understood now.

nazar-pc commented on domains/client/domain-executor/src/domain_block_processor.rs:
```diff
@@ -182,6 +196,130 @@ where
+
+                Ok(ForkAwareBlockImports {
+                    initial_parent: (parent_header.hash(), *parent_header.number()),
+                    primary_imports: retracted.iter().cloned().rev().collect::<Vec<_>>(),
```
This is simpler, or you can just call `retracted.reverse()`two lines above to even reuse allocation (not sure if it is reused here):
```suggestion
                    primary_imports: retracted.into_iter().rev().collect(),
```

nazar-pc replied:
Also I don't really understand why you return retracted blocks, they are already imported, what is the point of processing them again? I thought secondary chain still maintains proper fork tree already.

liuchengxu replied:
With more experiments, I can confirm you are right these retracted blocks are already processed and in chain, processing them again is redundant.

nazar-pc commented on domains/client/domain-executor/src/domain_block_processor.rs:
```diff
@@ -182,6 +196,130 @@ where
+        );
+
+        match (retracted.is_empty(), enacted.is_empty()) {
+            (true, false) => {
```
With below comments, all branches without `unreachable` when corrected, will end up being exactly the same (`common_block_number` thing will have to be done in all cases instead of `.info()`) unless I read it incorrectly. Which means we probably don't need `match` here at all.

nazar-pc replied:
@liuchengxu so with refactoring that you did we have 5 branches, one of which returns `None` (though could have returned an empty vector of pending primary block imports with the same result), two are unreachable and two remaining only differ in `initial_parent` construction. Moreover, I think second `initial_parent` construction is a super-set of the first one. What about removing `match` now and just use the last version as a universal return value for much simplified code?

liuchengxu replied:
It's intentional to return `None` as an empty vector is not semantically equivalent to a `None`. I usually would vote for simple code, but for this complicated topic, I'd really love to keep it as is for now so that each possible case can be handled/well explained, in case there is any misunderstanding somewhere, which will able to be found earlier at the early testing stage. not aesthetical, but practical IMHO, I can mark a TODO to simplify it in the future once enough confidence can be achieved though.

nazar-pc replied:
Well, the only thing we care here is what blocks from primary chain were not yet processed on a domain relatively to some parent. The rest IMHO is an implementation detail, so I don't see how `Option<Vec<T>>` is helping here.

nazar-pc reviewed (commented):
Mostly minor nits, but I feel like the major change can be greatly simplified

liuchengxu committed:
Apply a few review suggestions and improve some logs

- Remove unused code in subspace-fraud-proof tests
- Improve some logs to show more info in the domain block building
  pipeline
- Remove the redundant primary block processing in case `retracted` is
  non-empty and `enacted` is empty
- Rename `ForkAwareBlockImports` to `PendingPrimaryBlocks`, and add some
  docs.

liuchengxu reviewed (commented):
Testing locally looks like no problem so far, the next step will be run it on gemini-3b.

vedhavyas commented on domains/client/domain-executor/src/domain_worker.rs:
```diff
@@ -244,8 +236,8 @@ where
 async fn process_primary_block<Block, PBlock, PClient, ProcessorFn>(
-    domain_id: DomainId,
-    primary_chain_client: &PClient,
+    _domain_id: DomainId,
```
do we need the args if we are ignoring them ?

liuchengxu replied:
The whole function was eliminated in the later commits.

vedhavyas reviewed (commented):
Make sense. Will do another round of pending_primary_blocks in sometime

vedhavyas reviewed (approved):
Make sense. 
I dont see the pruning of already imported blocks that are now considered fork blocks. Could  this be the issue we are seeing ?

nazar-pc reviewed (approved):
Not opposed strongly, but I do feel like extra complexity is unnecessary here.

rahulksnv commented on domains/client/domain-executor/src/domain_block_processor.rs:
```diff
@@ -101,6 +201,123 @@ where
+        let route =
+            sp_blockchain::tree_route(&*self.primary_chain_client, primary_from, primary_to)?;
+
+        let retracted = route.retracted();
```
I don't get what retracted/enacted blocks mean. The documentation here https://paritytech.github.io/substrate/master/sp_blockchain/struct.TreeRoute.html doesn't explain much either. Can you please point me to some info on this?

nazar-pc replied:
It is just a path across forks over common for both branches block C (in diagram from above link).

liuchengxu committed:
Merge branch 'main' into fork-aware-process-bundles
